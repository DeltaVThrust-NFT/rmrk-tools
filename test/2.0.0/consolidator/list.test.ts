import { Consolidator } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  getAliceKey,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock,
  mintNftMock2,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";

beforeAll(async () => {
  return await cryptoWaitReady();
});

describe("rmrk2.0.0 Consolidator: LIST", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
    ...getBlockCallsMock(mintNftMock().mint()),
  ];

  it("Should allow to LIST listed NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock(3).list(BigInt(1e12))),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should prevent to LIST if parent is forsale", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock2().mint(mintNftMock(3).getId())),
      ...getBlockCallsMock(mintNftMock(3).list(BigInt(1e12))),
      ...getBlockCallsMock(mintNftMock2(4).list(BigInt(1e12))),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should prevent to LIST non-existent NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(mintNftMock(3).list(BigInt(1e12))),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should prevent to LIST burned NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock(3).burn()),
      ...getBlockCallsMock(mintNftMock(3).list(BigInt(1e12))),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should prevent to LIST non-owned NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(
        mintNftMock(3).list(BigInt(1e12)),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should prevent to LIST NFT owned by another NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock2().mint(mintNftMock(3).getId())),
      ...getBlockCallsMock(mintNftMock2(4).list(BigInt(1e12))),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });
});
