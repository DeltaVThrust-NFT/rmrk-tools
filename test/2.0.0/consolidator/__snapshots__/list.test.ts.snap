// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: LIST Should allow to LIST listed NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "attributes": Array [],
      "block": 3,
      "burned": "",
      "changes": Array [
        Object {
          "block": 4,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "forsale",
          "new": 1000000000000n,
          "old": 0n,
          "opType": "LIST",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 1000000000000n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: LIST Should prevent to LIST NFT owned by another NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 5,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[LIST] Attempting to list NFT 4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888 who's owner is another NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777.",
      "object_id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "op_type": "LIST",
    },
  ],
  "nfts": Array [
    Object {
      "attributes": Array [],
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [
        Object {
          "equipped": "",
          "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
          "pending": false,
        },
      ],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "attributes": Array [],
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "pending": false,
      "priority": Array [],
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: LIST Should prevent to LIST burned NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 5,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[LIST] Attempting to list burned NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "op_type": "LIST",
    },
  ],
  "nfts": Array [
    Object {
      "attributes": Array [],
      "block": 3,
      "burned": "true",
      "changes": Array [
        Object {
          "block": 4,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "burned",
          "new": "true",
          "old": "",
          "opType": "BURN",
        },
        Object {
          "block": 4,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "forsale",
          "new": 0n,
          "old": 0n,
          "opType": "BURN",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: LIST Should prevent to LIST non-existent NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [],
  "invalid": Array [
    Object {
      "block": 2,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[LIST] Attempting to list non-existant NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "op_type": "LIST",
    },
  ],
  "nfts": Array [],
}
`;

exports[`rmrk2.0.0 Consolidator: LIST Should prevent to LIST non-owned NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 4,
      "caller": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "message": "[LIST] Attempting to list non-owned NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777, real owner: 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "op_type": "LIST",
    },
  ],
  "nfts": Array [
    Object {
      "attributes": Array [],
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;
